src/sentry/issues/issue_velocity.py

```diff
@@ -0,0 +1,131 @@
+            Function("min", [Column("timestamp")], "first_seen"),  # when the issue was first seen
+            Function(
+                "countIf",
+                [
+                    Function("greaterOrEquals", [Column("timestamp"), one_week_ago])
+                ],  # count events for the issue that occurred within the past week
+                "past_week_event_count",
+            ),
+            Function(
+                "if",
+                [
+                    Function(
+                        "less", [Column("first_seen"), one_week_ago]
+                    ),  # if the issue is older than a week
+                    Function(
+                        "divide", [Column("past_week_event_count"), WEEK_IN_HOURS]
+                    ),  # divide the number of events in the week by a week in hours
+                    Function(
+                        "divide",
+                        [
+                            Column("past_week_event_count"),
+                            Function(
+                                "dateDiff", ["hour", Column("first_seen"), now]
+                            ),  # otherwise divide by its age in hours
+                        ],
+                    ),
+                ],
+                "hourly_event_rate",
+            ),
+        ],
+        groupby=[Column("group_id")],
+        where=[
+            Condition(
+                Column("timestamp"), Op.GTE, ninety_days_ago
+            ),  # include issues up to the oldest retention date to determine whether an issue is older than the week or not
+            Condition(Column("timestamp"), Op.LT, now),
+            Condition(Column("project_id"), Op.EQ, project.id),
+        ],
+        having=[
+            Condition(Column("past_week_event_count"), Op.GT, 1)
+        ],  # exclude any issues that had only 1 event in the past week
+        orderby=[OrderBy(Column("first_seen"), Direction.ASC)],
+        limit=Limit(10000),
+    )
+
+    query = Query(
+        match=subquery,
+        select=[
+            Function("quantile(0.9)", [Column("hourly_event_rate")], "p90")
+        ],  # get the approximate 90th percentile of the event frequency in the past week
+        limit=Limit(1),
+    )
+
+    request = Request(
+        dataset=Dataset.IssuePlatform.value,
+        app_id=REFERRER,
+        query=query,
+        tenant_ids={"referrer": REFERRER, "organization_id": project.organization.id},
+    )
+
+    result = raw_snql_query(request, referrer=REFERRER)["data"]
+    if len(result) == 0:
+        return None
+
+    try:
+        return result[0]["p90"]
+    except KeyError:
+        return None
+
+
+def set_velocity_threshold_for_project(project: Project) -> None:
+    threshold = calculate_velocity_threshold_for_project(project)
+    if threshold is None:
+        logger.error("Velocity threshold couldn't be calculated", extra={"project_id": project.id})
+        return
+    elif math.isnan(threshold):
+        pass  # TODO
```

Comment:
when we get NaN, should we refrain from storing in redis? this indicates that there were no events that met our conditions to calculate the threshold

post process job will need to be updated to handle when there is no threshold (i guess it'll just exit early when there is none)

Reply:
@isabellaenriquez if we get NaN, maybe we set the threshold to -1 to denote that we won't allow escalations? Or something else that says please don't try to escalate.

Reply:
oh i like the -1 idea, i think dan's pr is going to merge first so we can do the case handling in the pr where his dummy function is replaced